### PR TITLE
RFC: Use default retry mechanisms for Google Cloud APIs

### PR DESCRIPTION
**Pull request by bergman**
The Datastore operations should be good but the Bigtable operations are very unclear to me how to configure.

### CODE REVIEW

File: styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java [hunk 1]
```diff
@@ -177,31 +159,6 @@ private SequenceEvent parseEventResult(Result r) throws IOException {
...
 
-  private void storeWithRetries(RunnableWithException<IOException> storingOperation) throws IOException {
-    int storeRetries = 0;
-    boolean succeeded = false;
-
-    while (storeRetries < MAX_BIGTABLE_RETRIES && !succeeded) {
-      try {
-        storingOperation.run();
-        succeeded = true;
-      } catch (ResourceNotFoundException e) {
```
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java]: This doesn't throw this anymore so callers have to see for themselves if `e.getCause()` is a `ResourceNotFoundException`. Kind of clunky.

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java [hunk 2]
```diff
@@ -450,36 +434,6 @@ public WorkflowState workflowState(WorkflowId workflowId) throws IOException {
...
-    int storeRetries = 0;
-
-    while (storeRetries < MAX_RETRIES) {
-      try {
-        return storingOperation.apply();
-      } catch (ResourceNotFoundException e) {
-        throw e;
-      } catch (DatastoreException | IOException e) {
-        if (e.getCause() instanceof ResourceNotFoundException) {
-          throw (ResourceNotFoundException) e.getCause();
```
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: As with the Bigtable case, this check would have to live in the caller. Unless we stop using `runInTransaction` but I guess there is a reason we use that?

Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: I think we have more than one interactions for those cases, so wrapping inside a transaction makes sense.

File: styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java [hunk 3]
```diff
@@ -93,18 +81,16 @@
...
-        final String keyString = String.format("%s#%08d", workflowInstanceKey, sequenceEvent.counter());
-        final byte[] key = Bytes.toBytes(keyString);
-        final Put put = new Put(key, sequenceEvent.timestamp());
-
-        final byte[] eventBytes = serialize(sequenceEvent.event()).toByteArray();
-        put.addColumn(EVENT_CF, EVENT_QUALIFIER, eventBytes);
-        eventsTable.put(put);
-      }
-    });
+    final String workflowInstanceKey = sequenceEvent.event().workflowInstance().toKey();
```
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java]: Does it mean we won't retry for Bigtable?

File: styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java [hunk 4]
```diff
@@ -153,11 +139,7 @@ void writeEvent(SequenceEvent sequenceEvent) throws IOException {
       throws IOException {
     final Set<SequenceEvent> storedEvents = readEvents(workflowInstance);
     final Optional<SequenceEvent> lastStoredEvent = storedEvents.stream().reduce((a, b) -> b);
```
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java]: This way of getting the last element feels quite brutal. ;)

Maybe something like `return Optional.ofNullable(Iterators.getLast(storedEvents.iterator(), null)).map(SequenceEvent::counter);`?

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java (hunk 2 above)
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Hope `ResourceNotFoundException` will never be the cause of `DatastoreException` which is unfortunately a `RuntimeException`.

File: styx-common/src/main/java/com/spotify/styx/util/Connections.java [hunk 5]
```diff
@@ -54,6 +56,7 @@ public static Datastore createDatastore(Config config) {
     return DatastoreOptions.builder()
         .namespace(namespace)
         .projectId(projectId)
+        .retryParams(RetryParams.defaultInstance())
```
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/util/Connections.java]: This seems to be the default behaviour which means by default it retries. Check `ServiceOptions` L217. We have been doing something on top of the default retry.

Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/util/Connections.java]: You're right!

File: styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java (hunk 4 above)
Comment (bergman) [inline styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java]: 👍

**Comment by honnix** [comment]
I think we agreed to park this PR for now and upgrade datastore client lib first and see whether it solves retry problem. Am I right?

**Comment by honnix** [comment]
Shall we close this? Application layer retry seems still appreciated.

**Pull request closed without merging**